design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+
+## Design
+
+The problem at hand is how to determine if a server is down and where to route a request when it is 
+down. Every KSQL server must have available the information of what other KSQL servers exist in the 
+cluster, their status and their lag per topic partition. There are three components to this: 
+Healthchecking, Lag reporting and Lag-aware routing.  
+
+### Failure detection/Healthcheck
+
+Failure detection/ healthcheck is implemented via a periodic hearbeat. KSQL servers either broadcast 
+their heartbeat (N^2 interconnections with N KSQL servers) or we implement a gossip protocol. In 
+the initial implementation, we will use the REST API to send heartbeats leveraging the N^2 mesh that 
+already exists between KSQL servers. Hence, we will implement a new REST endpoint, `/heartbeat` that 
+will register the heartbeats a server receives from other servers. 
+
+We want to guarantee 99% uptime. This means that in a 5 minute window, it must take no longer than 
+2 seconds to determine the active server is down and to forward the request to one of the standbys. 
+The heartbeats must be light-weight so that we can send multiple heartbeats per second which will 
+provide us with more datapoints required to implement a well-informed policy for determining when a 
+server is up and when down.
+
+Configuration parameters:
+1) Heartbeat INTERVAL (e.g send heartbeat every 200 ms)
+2) Heartbeat WINDOW (e.g every 2 seconds, count the missed/received heartbeats and determine if server is down/up)
+3) MISSED_THRESHOLD: How many heartbeats in a row constitute a node as down (e.g. 3 missed heartbeats = server is down)
+4) RECEIVED_THRESHOLD: How many heartbeats in a row constitute a node as up (e.g. 2 received heartbeats = server is up)
+
+We will provide sane defaults out-of-box, to achieve a 99% uptime.
+
+Pseudocode for REST endpoint for `/heartbeat`:
+```java
+Map<KsqlServer, List<Long>> receivedHeartbeats;
+
+@POST
+public Response receiveHeartbeat(Request heartbeat) 
+  return Response.ok(processRequest()).build();
+}
+
+private HeartbeatResponse processRequest() {
+  KsqlServer server = request.getServer();
+  receivedHeartbeats.get(server).add(System.currentTimeMillis());
+}
+```
+
+
+### Lag reporting
+
+Every server neeeds to periodically broadcast their local lag information. We will implement a new 
+REST endpoint `/reportlag`. The lag information at a server is obtained via 
+`Map<String, Map<Integer, Long>> localLagMap = kafkaStreams.allLocalOffsetLags();`.
+
+Pseudocode for REST endpoint for `/reportlag`:
+```java
+Map<KsqlServer, Map<String, Map<Integer, Long>> globalLagMap;
+
+@POST
+public Response receiveLagInfo(Request lagMap) 
+  return Response.ok(processRequest()).build();
+}
+
+private LagReportResponse processRequest() {
+  KsqlServer server = request.getServer();
+  Map<String, Map<Integer, Long>> localLagMap = request.getLag();
+  globalLagMap.put(server, localLagMap);
+}
+``` 
+
+### Lag-aware routing
```

Comment:
I am not sure what you mean by "hot nodes" as either way, a query is served by one node only, the one that has the key. We cannot round robin among the standbys because different queries for the same key will return different results based on the lag of each standby. Hence, there is no consistency guarantee until a new active is elected. However, if we forward to the server with the smallest lag, we have best effort consistency guarantee. 
What is v1 and how does it get simplified?

Reply:
I guess I was trying to understand our consistency model. As long as we're willing to serve from stale replicas, then we will not be consistent anyway (and can't market ksqlDB as such) - at that point, why bother serving from just the most up-to-date node? What's a realistic lag in the scenario of a master failover?

v1 I just meant the first version of HA queries - and it becomes simpler because we wouldn't need to consider lag-aware routing at all.

Reply:
There are two things: 1. Eventual consistency and 2. Determinism. We have eventual consistency by means of serving stale data until active is up or lag is zero. Moreover, we offer guarantees of serving the most up-to-date data. We are also deterministic by means of returning the same result for the same query (if nothing changes in the system). With what you propose we are not deterministic as queries return random results based on which server serves the query. This is a matter of correctness and not consistency the way I see it. Moreover, there is no guarantees on the consistency either which is also not sound.

Reply:
> We are also deterministic by means of returning the same result for the same query (if nothing changes in the system)

"If something changes in the system" is exactly the situation we're discussing- you have no determinism if the master goes down (you could get a different result for the same query), so why is an artificial level of determinism better? What I'm saying is that you can't actually give any better of a hard SLA with this proposal as you can with just round-robin. It's a "best-effort" either way.

Reply:
>> why bother serving from just the most up-to-date node? 
A system sacrificing C, does not mean its not ever consistent.. During normal mode of operations, if you keep reading from active replica, you can get consistent results which has significant value and actually makes this whole eventual consistency model tolerable. 

>> What's a realistic lag in the scenario of a master failover?
We need to benchmark and understand.. This is an under explored area for Kafka streams in general. 

Lag aware routing solves a basic usability problem in the initial version. If the standby's are lagging way way behind or even if the active is restoring way way behind, its probably safe to fail queries than to leak some very stale, incorrect data..

Reply:
I think was @agavra tries to say is, if there are two standbys, and both are up-to-date enough to meet the "max lag" criteria, why not use both to server queries if the active is not available to get better load balancing. (@agavra -- please correct me if I miss understood). This seems to be a fair idea in general, as we serve stall data anyway if the active is not available and the only guarantee we provide is "max lag" that both standbys could fulfill.

However, I don't think it would make v1 simpler but more complex, because we would need to have some measure of "load" to decide which of both standbys should answers the query. A simple round robin might not help much given the fact that we don't know how state store shards are distributed and which shards might be queried. Assuming there are enough shards we can "safely" assume (or hope) for a natural load balancing. If this assumption does not hold up, I doubt that round robin would help, as it would most likely just move a hotspot to another server?